spyder_kernels/comms/frontendcomm.py

```diff
@@ -127,22 +127,52 @@ def cache_message(self, comm_id, msg):
 
     def notify_comm_ready(self, comm):
         """Send messages about comm readiness to frontend."""
-        self.remote_call(comm.comm_id)._comm_ready()
+        self.remote_call(
+            comm_id=comm.comm_id,
+            callback=self._comm_ready_callback
+        )._comm_ready()
 
         # Cached messages for that comm
         if comm.comm_id in self._cached_messages:
             for msg in self._cached_messages[comm.comm_id]:
                 comm.handle_msg(msg)
             self._cached_messages.pop(comm.comm_id)
 
-    def notify_iopub_ready(self):
-        """Notify frontend that the iopub is ready"""
-        self._iopub_ready = True
-        for comm in self._pending_comms:
-            self.notify_comm_ready(comm)
-        self._pending_comms = []
-
     # --- Private --------
+    def _wait_for_iopub_connect(self, comm):
+        """Wait 10 seconds for iopub new connections."""
+        # This is necessary because any iopub message 
+        # (including comm responses) will be dropped before iopub connection is
+        # finished. (See https://zguide.zeromq.org/docs/chapter1 :
+        # ** the subscriber will always miss the first messages that the 
+        #    publisher sends)
+        if self.iopub_monitor is None:
+            self.iopub_monitor = self.kernel.iopub_socket.get_monitor_socket(
+                zmq.EVENT_HANDSHAKE_SUCCEEDED)
+        iopub_monitor_thread = threading.Thread(
+            target=self._iopub_monitor_main, args=(comm.comm_id, ))
+        self._pending_comms[comm.comm_id] = (comm, iopub_monitor_thread)
+        iopub_monitor_thread.start()
+
+    def _iopub_monitor_main(self, comm_id):
+        """
+        Send comm message to frontend when comms are connected
+        """
+        if self.iopub_monitor.poll(1e4):
+            # If a message is recieved on iopub_monitor, there is a good chance
+            # That it will go through. Try sending again if it passed
```

Comment:
```suggestion
            # If a message is received on iopub_monitor, there is a good chance
            # that it will go through. Try sending it again if it passed.
```